tests/framework/scheduler.py

```diff
@@ -225,40 +291,69 @@ def pytest_sessionfinish(self, session, exitstatus):
-
-        Called in the server process context.
+    def wait_for_worker_to_die(self, workers):
         """
-        max_workers = batch['concurrency']
-        items_per_worker = max(1, int(len(batch['items']) / max_workers))
-        workers = []
-        while batch['items']:
-            # Pop `items_per_worker` out from this batch and send them to
-            # a new worker.
-            worker_items = batch['items'][-items_per_worker:]
-            del batch['items'][-items_per_worker:]
-
-            # Avoid storming the host with too many workers started at once.
-            _delay = random() + len(workers) / 5.0 if max_workers > 1 else 0
-
-            # Create the worker process and start it up.
-            worker = mp.Process(
-                target=self._worker_main,
-                args=(worker_items, _delay)
-            )
-            workers.append(worker)
-            worker.start()
+        Wait for a worker to die.
 
-        # Main loop, reaping workers and processing IPC requests.
-        while workers:
+        Polls worker list and handles any IPC calls.
+        If one of the workers dies, it removes it from the workers dict.
+        """
+        while self.test_running:
             rlist, _, _ = select(self._mp_singletons, [], [], 0.1)
             for mps in rlist:
                 mps.handle_ipc_call()
-            _ = [w.join() for w in workers if not w.is_alive()]
-            workers = [w for w in workers if w.is_alive()]
 
-    def _worker_main(self, items, startup_delay=0):
+            for worker in list(workers.keys()):
+                if not worker.is_alive():
+                    del workers[worker]
+
+    def _run_items(self, items):
+        """Run test items and spread them by concurrency."""
+        max_concurrency = self.session.config.option.concurrency
+        test_items = [
+            PytestScheduler.TestItem(i, max_concurrency) for i in items
+        ]
+
+        # Sort test items by highest concurrency value first
+        items_prio = sorted(
+            test_items, key=lambda c: c.concurrency, reverse=True
+        )
+
+        workers = {}
+        self.test_running = True
+        # Start a thread that reaps workers
+        watcher = threading.Thread(
+            target=self.wait_for_worker_to_die, args=(workers,))
+        watcher.start()
+
+        while items_prio:
+            # Pop items and start processes as long as something can run
+            if items_prio[0].can_run(workers):
+                crt_item = items_prio.pop(0)
+
+                # Avoid storming the host with too many workers
+                # started at once.
+                _delay = (
+                    random() + len(workers) / 5.0 if len(workers) > 1 else 0
+                )
+
+                # Create the worker process and start it up.
+                worker = mp.Process(
+                    target=self._worker_main, args=(crt_item.item, _delay)
```

Comment:
it seems that we are using a worker for only one item, right? Can't we give a worker more items to handle sequentially?